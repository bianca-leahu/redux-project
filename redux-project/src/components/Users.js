import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Form, FormGroup, Label, Input, Button, FormFeedback, Alert } from 'reactstrap';

export default class Users extends Component {
    static propTypes = {
        users: PropTypes.array.isRequired,
        projects: PropTypes.array.isRequired,
        roles: PropTypes.array.isRequired,
        showForm: PropTypes.func.isRequired,
        isFormOpen: PropTypes.bool.isRequired,
        handleChange: PropTypes.func.isRequired,
        closeForm: PropTypes.func.isRequired,
        userName: PropTypes.string.isRequired,
        noProjects: PropTypes.bool.isRequired,
        noRoles: PropTypes.bool.isRequired,
        showErrorMessage: PropTypes.bool.isRequired,
        hasError: PropTypes.bool.isRequired,
        handleSubmit: PropTypes.func.isRequired
    }

    /**
     * Handle the user details form submission
     * @param {Object} e - the "click" event object
     */
    handleSubmit = (e) => {
        e.preventDefault();

        this.props.handleSubmit(this.projectsInput.value, this.rolesInput.value)
    }

    render() {
    	const { users, projects, roles, showForm, isFormOpen, handleChange, closeForm, userName, noProjects, noRoles, showErrorMessage, hasError } = this.props;

        return (
            <div className="users-page">
    			{!hasError 
                    ? <div className="users-page__list">
        				<h1 className='my-4 text-center'>Users</h1>
          				<div>
        					{users.map(item => (
          						<p className={`mt-2 mb-4 users-page__list-item ${item.name === userName && isFormOpen ? 'is-active' : ''}`} key={item.id} onClick={() => showForm(item.id, item.name, item.roles, item.projects)}>
            						{item.name} <br />
                                    <small>{item.roles ? `is assigned ${item.roles.toUpperCase()} role` : ''}</small>
                                    {item.roles ? <br /> : null}
                                    <small>{item.projects ? `in the ${item.projects.toUpperCase()} project` : ''}</small>
          						</p>
        					))}
      					</div>
        			</div>
                    :  <Alert color="danger">
                        This is a danger alert â€” check it out!
                    </Alert>}

    			{isFormOpen ? <Form onSubmit={this.handleSubmit} className='border users-page__form'>
    				<FormGroup>
                        <h2>{userName}</h2>
                        <p className='users-page__close' onClick={closeForm}>X</p>
          				<Label for="projects">Projects</Label>
          				<Input type="select" name="select" id="projects" invalid={noProjects} innerRef={(node) => this.projectsInput = node} onChange={(e) => handleChange(e)}>
            				<option value=''>Choose a project</option>
				            {projects.map(item => (
	      						<option key={item.id} value={item.id}>{item.name}</option>
    						))}
          				</Input>
          				<FormFeedback>You need to choose a project</FormFeedback>
        			</FormGroup>

        			<FormGroup>
          				<Label for="roles">Roles</Label>
          				<Input type="select" name="select" id="roles" invalid={noRoles} innerRef={(node) => this.rolesInput = node} onChange={(e) => handleChange(e)}>
            				<option value=''>Choose a role</option>
				            {roles.map(item => (
	      						<option key={item.id} value={item.id}>{item.name}</option>
    						))}
          				</Input>
          				<FormFeedback>You need to choose a role</FormFeedback>
        			</FormGroup>

                    {showErrorMessage === true ? <span className='d-flex mb-4 text-danger'><small>This role for this project has already been taken. Please choose a different role or a diferent project</small></span> : null}

        			<Button>Submit</Button>
    			</Form> : null}
  			</div>
        );
    }
}