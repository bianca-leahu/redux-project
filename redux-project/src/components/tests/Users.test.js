import React from 'react';
import renderer from 'react-test-renderer';
import { configure, shallow } from 'enzyme';
import Users from '../Users';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() })

const users = [
	{
		id: 1,
        name: 'John Doe',
    }
];

const projects = [
	{
		id: 1,
		name: 'Project',
	}
];

const roles = [
	{
		id: 1,
		name: 'Role',
	}
]

const usersProps = {
    users,
    projects,
    roles,
    showForm: jest.fn,
    isFormOpen: false,
    handleChange: jest.fn,
    closeForm: jest.fn,
    userName: '',
    noProjects: false,
    noRoles: false,
    showErrorMessage: false,
    hasError: false,
    handleSubmit: jest.fn
};

describe('<Users />', () => {
	it('should render Users', () => {
		const wrapper = renderer.create(<Users  {...usersProps}/>)
	})
	it('matches the snapshot', () => {
		const tree = shallow(<Users  {...usersProps}/>)
		expect(tree).toMatchSnapshot()
	})
});