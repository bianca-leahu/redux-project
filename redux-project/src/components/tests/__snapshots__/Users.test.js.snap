// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Users
    closeForm={[Function]}
    handleChange={[Function]}
    handleSubmit={[Function]}
    hasError={false}
    isFormOpen={false}
    noProjects={false}
    noRoles={false}
    projects={
      Array [
        Object {
          "id": 1,
          "name": "Project",
        },
      ]
    }
    roles={
      Array [
        Object {
          "id": 1,
          "name": "Role",
        },
      ]
    }
    showErrorMessage={false}
    showForm={[Function]}
    userName=""
    users={
      Array [
        Object {
          "id": 1,
          "name": "John Doe",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="users-page__list"
        >
          <h1
            className="my-4 text-center"
          >
            Users
          </h1>
          <div>
            <p
              className="mt-2 mb-4 users-page__list-item "
              onClick={[Function]}
            >
              John Doe
               
              <br />
              <small>
                
              </small>
              <small>
                
              </small>
            </p>
          </div>
        </div>,
        null,
      ],
      "className": "users-page",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="my-4 text-center"
            >
              Users
            </h1>,
            <div>
              <p
                className="mt-2 mb-4 users-page__list-item "
                onClick={[Function]}
              >
                John Doe
                 
                <br />
                <small>
                  
                </small>
                <small>
                  
                </small>
              </p>
            </div>,
          ],
          "className": "users-page__list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Users",
              "className": "my-4 text-center",
            },
            "ref": null,
            "rendered": "Users",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="mt-2 mb-4 users-page__list-item "
                  onClick={[Function]}
                >
                  John Doe
                   
                  <br />
                  <small>
                    
                  </small>
                  <small>
                    
                  </small>
                </p>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "John Doe",
                    " ",
                    <br />,
                    <small>
                      
                    </small>,
                    null,
                    <small>
                      
                    </small>,
                  ],
                  "className": "mt-2 mb-4 users-page__list-item ",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Array [
                  "John Doe",
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {},
                    "ref": null,
                    "rendered": null,
                    "type": "br",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "",
                    },
                    "ref": null,
                    "rendered": "",
                    "type": "small",
                  },
                  null,
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "",
                    },
                    "ref": null,
                    "rendered": "",
                    "type": "small",
                  },
                ],
                "type": "p",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="users-page__list"
          >
            <h1
              className="my-4 text-center"
            >
              Users
            </h1>
            <div>
              <p
                className="mt-2 mb-4 users-page__list-item "
                onClick={[Function]}
              >
                John Doe
                 
                <br />
                <small>
                  
                </small>
                <small>
                  
                </small>
              </p>
            </div>
          </div>,
          null,
        ],
        "className": "users-page",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="my-4 text-center"
              >
                Users
              </h1>,
              <div>
                <p
                  className="mt-2 mb-4 users-page__list-item "
                  onClick={[Function]}
                >
                  John Doe
                   
                  <br />
                  <small>
                    
                  </small>
                  <small>
                    
                  </small>
                </p>
              </div>,
            ],
            "className": "users-page__list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Users",
                "className": "my-4 text-center",
              },
              "ref": null,
              "rendered": "Users",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="mt-2 mb-4 users-page__list-item "
                    onClick={[Function]}
                  >
                    John Doe
                     
                    <br />
                    <small>
                      
                    </small>
                    <small>
                      
                    </small>
                  </p>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      "John Doe",
                      " ",
                      <br />,
                      <small>
                        
                      </small>,
                      null,
                      <small>
                        
                      </small>,
                    ],
                    "className": "mt-2 mb-4 users-page__list-item ",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": Array [
                    "John Doe",
                    " ",
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {},
                      "ref": null,
                      "rendered": null,
                      "type": "br",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "",
                      },
                      "ref": null,
                      "rendered": "",
                      "type": "small",
                    },
                    null,
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "",
                      },
                      "ref": null,
                      "rendered": "",
                      "type": "small",
                    },
                  ],
                  "type": "p",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
